core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -250,6 +250,9 @@ public void run() {
       loadBalancer.shutdown();
       loadBalancer = null;
```

Comment:
If the factory is shared among multiple channels, it's not worse as long as the factory only recycles the completely shutdown ones.

What about letting `ManagedChannelImpl` have a `helper` field instead of `loadBalancer` field. And in `LbHelperImpl`, checking `this == ManagedChannelImpl.this.helper` instead of checking `loadBalancer == lb`?